### PR TITLE
Documentation for loss functions

### LINKED ISSUE #3480: IClassificationLoss implementations should be documented
There are a few implementations of `IClassificationLoss`, most have them have little or no documentation. For example, the documentation of `ExpLoss` says:
Exponential Loss

### PR DESCRIPTION
**Pull request by najeeb-kazmi**
Fixes #3480

### CODE REVIEW

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 1]
```diff
@@ -222,19 +262,46 @@ public float DualUpdate(float output, float label, float alpha, float invariant,
...
+    /// Let $L(f(\vec{x}, y) = 1 - yf(\vec{x})$, where $f(\vec{x})$ is the predicted score and $y \in \{-1, 1\}$ is the true label.
+    ///
+    /// Note that the labels used in this calculation are -1 and 1, unlike Log Loss, where the labels used are 0 and 1.
+    /// Also unlike Log Loss, $f(\vec{x})$ is the raw predicted score, not the predicted probability (which is calculated by applying a [sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function) to the predicted score).
+    ///
+    /// The Smoothed Hinge Loss function is then defined as:
+    ///
+    /// \[
+    /// g(L(f(\vec{ x}, y)) =
+    /// \begin{cases}
```
Comment (najeeb-kazmi) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: @natke @sfilipi do you know if this will render? This is using the amsmath LaTeX package.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 2]
```diff
@@ -445,8 +554,26 @@ public float Derivative(float output, float label)
...
-    /// Tweedie loss, based on the log-likelihood of the Tweedie distribution.
+    /// Tweedie loss, based on the log-likelihood of the Tweedie distribution. This loss function is used in Tweedie regression.
     /// </summary>
+    /// <remarks type="text/markdown"><![CDATA[
+    ///
+    /// The Tweedie Loss function is defined as:
+    ///
+    /// \[
+    /// L(f(\vec{ x}, y, i) =
+    /// \begin{cases}
```
Comment (najeeb-kazmi) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: @natke @sfilipi do you know if this will render? This is using the amsmath LaTeX package.

Comment (najeeb-kazmi) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: @natke @sfilipi do you know if this will render? This is using the amsmath LaTeX package.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 1 above)
Comment (najeeb-kazmi) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: @natke @sfilipi do you know if this will render? This is using the amsmath LaTeX package.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 3]
```diff
@@ -31,7 +31,7 @@
 [assembly: LoadableClass(TweedieLoss.Summary, typeof(TweedieLoss), typeof(TweedieLoss.Options), typeof(SignatureRegressionLoss),
     "Tweedie Loss", "TweedieLoss", "Tweedie", "Tw")]
 
-[assembly: EntryPointModule(typeof(ExpLoss.Options))]
+[assembly: EntryPointModule(typeof(ExponentialLoss.Options))]
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >ExponentialLoss [](start = 35, length = 15)

something tells me you need to run regenerateEntryPoints test :)

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 4]
```diff
@@ -105,12 +105,32 @@ internal sealed class LogLossFactory : ISupportSdcaClassificationLossFactory, IS
...
+    ///
+    /// The Log Loss function is defined as:
+    ///
+    /// $L(f(\vec{x}, y) = -y ln(f(\vec{x})) - (1 - y) ln(1 - f(\vec{x}))$
+    ///
+    /// where $f(\vec{x})$ is the predicted probability of belonging to the positive class and $y \in \{0, 1\}$ is the true label.
+    ///
+    /// Note that the labels used in this calculation are 0 and 1, unlike Hinge Loss and Exponential Loss, where the labels used are -1 and 1.
+    ///
+    /// The Log Loss function provides a measure of how "certain" a classifier's predictions are, instead of just measuring how "correct" they are.
```
Comment (Ivanidzo4ka) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >Log [](start = 12, length = 3)

you have kink on wikipedia hinge loss, any reason why you decide not add link for log loss?

Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: Hinge loss had some interesting details regarding differentiability, that's why I added the link. Log loss section on that article doesn't have much to add.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 3 above)
Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: Apparently not. I ran RegenerateEntryPointCatalog test and there was no diff in the json manifest or ep-list.tsv. Also the build and tests pass without running RegenerateEntryPointCatalog.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 2 above)
Comment (natke) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: I think it should be ok @najeeb-kazmi. The only thing we've had trouble with so far is \boldsymbol. We can test it before you merge this PR if you like

Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: @natke yes let's test it please before I merge. Let me know if there are any errors. Thanks!

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 1 above)
Comment (natke) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: Preview looks good: https://review.docs.microsoft.com/en-us/dotnet/api/microsoft.ml.trainers.logloss?branch=smoke-test-preview&view=ml-dotnet

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 2 above)
Comment (natke) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: Preview looks good: https://review.docs.microsoft.com/en-us/dotnet/api/microsoft.ml.trainers.logloss?branch=smoke-test-preview&view=ml-dotnet

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 1 above)
Comment (natke) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: This doesn't work. For a start, there is no $ $ surrounding the latex. You can test it by editing the smoke-test-preview branch here:  https://github.com/dotnet/ml-api-docs/blob/smoke-test-preview/dotnet/xml/Microsoft.ML.Trainers/SmoothedHingeLoss.xml

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 2 above)
Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: @natke my question was regarding SmoothedHingeLoss and TweedieLoss. I don't see the new documentation for those on the site.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 1 above)
Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: The \\[ and \\] should have worked the same as $ $. In any case, I've updated this PR to use $ $ instead.

I tried pushing a commit to the smoke-test-preview branch but got access denied.

**Comment by shmoradims** [comment]
> // See the LICENSE file in the project root for more information.

please revert this file. no-code changes with documentation work plz, even trivial ones like these.

---
Refers to: src/Microsoft.ML.Core/Utilities/MathUtils.cs:3 in 96f130c. [](commit_id = 96f130c33ae2f9f12b2642a503437f77095d3055, deletion_comment = False)

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 5]
```diff
@@ -319,48 +386,65 @@ public float DualUpdate(float output, float label, float alpha, float invariant,
...
+    ///
+    /// Note that the labels used in this calculation are -1 and 1, unlike Log Loss, where the labels used are 0 and 1.
+    /// Also unlike Log Loss, $f(\vec{x})$ is the raw predicted score, not the predicted probability (which is calculated by applying a [sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function) to the predicted score).
+    ///
+    /// The Exponential Loss function penalizes incorrect predictions more than the Hinge loss and has a larger gradient.
+    ///
+    /// ]]>
+    /// </format>
+    /// </remarks>
+    public sealed class ExponentialLoss : IClassificationLoss
```
Comment (shmoradims) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >ExponentialLoss [](start = 24, length = 15)

i prefer code-changes being done in separate PRs. But I'm ok if Shauheen approves this change.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 6]
```diff
@@ -105,12 +105,32 @@ internal sealed class LogLossFactory : ISupportSdcaClassificationLossFactory, IS
...
+    /// The Log Loss, also known as the Cross Entropy Loss. It is commonly used in classification tasks.
+    /// </summary>
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// The Log Loss function is defined as:
+    ///
+    /// $L(f(\vec{x}), y) = -y ln(f(\vec{x})) - (1 - y) ln(1 - f(\vec{x}))$
+    ///
+    /// where $f(\vec{x})$ is the predicted probability of belonging to the positive class and $y \in \{0, 1\}$ is the true label.
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >f(\vec{x}) [](start = 15, length = 10)

Would it be better to give this a name? Maybe \hat(y)?

Comment (yaeldekel) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: And I would also explain that it is the sigmoid of the score.

---
In reply to: [278258057](https://github.com/dotnet/machinelearning/pull/3560#discussion_r278258057) [](ancestors = 278258057)

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 7]
```diff
@@ -141,28 +161,48 @@ public float DualUpdate(float output, float label, float dual, float invariant,
...
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// The Hinge Loss function is defined as:
+    ///
+    /// $L(f(\vec{x}), y) = max(0, m - yf(\vec{x}))$
+    ///
+    /// where $f(\vec{x})$ is the predicted score, $y \in \{-1, 1\}$ is the true label, and $m$ is the margin parameter set to 1 by default.
+    ///
+    /// Note that the labels used in this calculation are -1 and 1, unlike Log Loss, where the labels used are 0 and 1.
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >Log Loss [](start = 75, length = 8)

Use cref.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 8]
```diff
@@ -222,19 +262,46 @@ public float DualUpdate(float output, float label, float alpha, float invariant,
...
             if (dual < 0 || dual > 1)
-                return Double.NegativeInfinity;
+                return double.NegativeInfinity;
 
             return _margin * dual;
         }
     }
 
+    /// <summary>
+    /// A smooth version of the Hinge Loss function, commonly used in classification tasks.
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >Hinge Loss [](start = 32, length = 10)

cref.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 9]
```diff
@@ -222,19 +262,46 @@ public float DualUpdate(float output, float label, float alpha, float invariant,
...
+    /// </summary>
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// Let $L(f(\vec{x}), y) = 1 - yf(\vec{x})$, where $f(\vec{x})$ is the predicted score and $y \in \{-1, 1\}$ is the true label.
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >$L(f(\vec{x}), y) = 1 - yf(\vec{x})$ [](start = 12, length = 36)

Can you mention somewhere that this is the definition of the hinge loss? Maybe something like "Let $...$ denote the hinge loss".

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 10]
```diff
@@ -319,48 +386,65 @@ public float DualUpdate(float output, float label, float alpha, float invariant,
...
+    /// <remarks>
+    /// <format type="text/markdown"><![CDATA[
+    ///
+    /// The Exponential Loss function is defined as:
+    ///
+    /// $L(f(\vec{x}), y) = e^{-\beta y f(\vec{x})}$
+    ///
+    /// where $f(\vec{x})$ is the predicted score, $y \in {-1, 1}$ is the true label, and $\beta$ is a scale factor set to 1 by default.
+    ///
+    /// Note that the labels used in this calculation are -1 and 1, unlike Log Loss, where the labels used are 0 and 1.
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >Log Loss, [](start = 75, length = 9)

cref.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 11]
```diff
@@ -424,13 +523,23 @@ internal sealed class PoissonLossFactory : ISupportRegressionLossFactory
...
-    /// Poisson Loss.
+    /// Poisson Loss function for Poisson Regression.
     /// </summary>
+    /// <remarks type="text/markdown"><![CDATA[
+    ///
+    /// The Poisson Loss function is defined as:
+    ///
+    /// $L(f(\vec{x}), y) = e^{f(\vec{x})} - yf(\vec{x})$
+    ///
+    /// where $f(\vec{x})$ is the predicted value, $y$ is the true label.
```
Comment (yaeldekel) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >$f(\vec{x})$ [](start = 14, length = 12)

For all the loss functions, I think it would be easier to read if this expression had a name, such as \hat(y). In the log loss case, you can define the probability p as a function of \hat(y), and then use p in the formula for the loss.

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs (hunk 5 above)
Comment (yaeldekel) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: We should not change class names at this point.

---
In reply to: [278200927](https://github.com/dotnet/machinelearning/pull/3560#discussion_r278200927) [](ancestors = 278200927)

Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: We have been changing all abbreviated forms to the full names. I think the loss functions were overlooked because we were focusing on trainers and transforms. If there is a strong opinion that names should not change now, I will revert this change.

@shauheen @TomFinley @yaeldekel @glebuk

Comment (najeeb-kazmi) [reply src/Microsoft.ML.Data/Utils/LossFunctions.cs]: Reverting

File: src/Microsoft.ML.Data/Utils/LossFunctions.cs [hunk 12]
```diff
@@ -105,6 +105,26 @@ internal sealed class LogLossFactory : ISupportSdcaClassificationLossFactory, IS
...
+    ///
+    /// The Log Loss function is defined as:
+    ///
+    /// $L(p(\hat{y}), y) = -y ln(\hat{y}) - (1 - y) ln(1 - \hat{y})$
+    ///
+    /// where $\hat{y}$ is the predicted score, $p(\hat{y})$ is the probability of belonging to the positive class by applying a [sigmoid function](https://en.wikipedia.org/wiki/Sigmoid_function) to the score, and $y \in \\{0, 1\\}$ is the true label.
+    ///
+    /// Note that the labels used in this calculation are 0 and 1, unlike [Hinge Loss](xref:Microsoft.ML.Trainers.HingeLoss) and [Exponential Loss](xref:Microsoft.ML.Trainers.ExpLoss), where the labels used are -1 and 1.
+    ///
+    /// The Log Loss function provides a measure of how "certain" a classifier's predictions are, instead of just measuring how "correct" they are.
```
Comment (shmoradims) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >"certain" [](start = 56, length = 9)

*certain* (use italic instead of "")

Comment (shmoradims) [inline src/Microsoft.ML.Data/Utils/LossFunctions.cs]: >"correct" [](start = 128, length = 9)

ditto
